StreamChatClient_v3/Controllers/EntityDatabaseObserver.swift

```diff
@@ -0,0 +1,119 @@
+//
+
+import CoreData
+import Foundation
+
+/// This enum describes the changes to a certain item when observing it.
+public enum EntityChange<Item> {
+    /// The item was created or the recent changes to it make it match the predicate of the observer.
+    case create(_ item: Item)
+    
+    /// The item was updated.
+    case update(_ item: Item)
+    
+    /// The item was deleted or it no longer matches the predicate of the observer.
+    case remove(_ item: Item)
+}
+
+extension EntityChange: Equatable where Item: Equatable {}
+
+extension EntityChange {
+    /// Create a `EnitityChange` value from the provided `ListChange`. It simply transforms `ListChange` to `EntityChange`
+    /// be removing the cases that don't make sense for a single entity change, and using a better naming.
+    init(listChange: ListChange<Item>) {
+        switch listChange {
+        case .insert(let item, index: _):
+            self = .create(item)
+        case let .move(item, _, _), .update(let item, index: _):
+            self = .update(item)
+        case .remove(let item, index: _):
+            self = .remove(item)
+        }
+    }
+}
+
+/// Observes changes of a single entity specified using an `NSFetchRequest`in the provided `NSManagedObjectContext`.
+class EntityDatabaseObserver<Item, DTO: NSManagedObject> {
+    /// The observed item. `nil` of no item matches the predicate or the item was deleted.
+    @Cached var item: Item?
+    
+    /// Called with the aggregated changes after the internal `NSFetchResultsController` calls `controllerDidChangeContent`
+    /// on its delegate.
+    var onChange: ((EntityChange<Item>) -> Void)? {
+        didSet {
+            changeAggregator.onChange = { [unowned self] listChanges in
+                log.assert(listChanges.count <= 1, "EntityDatabaseObserver predicate shouldn't produce more than one change")
+                if let entityChange = listChanges.first.map(EntityChange.init) {
+                    self._item.reset()
+                    self.onChange?(entityChange)
+                }
+            }
+        }
+    }
+    
+    /// Acts like the `NSFetchedResultsController`'s delegate and aggregates the reported changes into easily consumable form.
+    private(set) lazy var changeAggregator: ListChangeAggregator<DTO, Item> =
+        ListChangeAggregator<DTO, Item>(itemCreator: self.itemCreator)
+    
+    /// Used for observing the changes in the DB.
+    private(set) lazy var frc: NSFetchedResultsController<DTO> = self.fetchedResultsControllerType
+        .init(fetchRequest: self.request,
+              managedObjectContext: self.context,
+              sectionNameKeyPath: nil,
+              cacheName: nil)
+    
+    /// The `NSFetchedResultsController` subclass the observe uses to create its FRC. You can inject your custom subclass
+    /// here if needed, i.e. when testing.
+    var fetchedResultsControllerType: NSFetchedResultsController<DTO>.Type = NSFetchedResultsController<DTO>.self
+    
```

Comment:
We have `fetchedResultsControllerType` here, but `frc` is `lazy`. 
So it's possible that type and actual `frc` will be different. 
Maybe it makes sense to reset `frc` after new type is set?

Reply:
Very good catch! Thanks

Reply:
Fixed in a9cef171afbd1c63fa0337e554462b1620ca5ea0 and 84de944bbf45c8fe60e553390535f5661b1a1899